### ISSUE TITLE
Missing perl-deprecate module not properly identified

### ISSUE THREAD

**Issue by saieva** [LMS-Community/slimserver#810]
If `perl-deprecate` is not installed `slimserver.pl` startup will fail with the message:

> The following CPAN modules were found but cannot work with Logitech Media Server:
>  CGI::Cookie (loaded <not found>, need 1.27)

Recommend adding `deprecate` to `@default_required_modules` in `Slim/bootstrap.pm` so that an accurate error message is produced for this issue.

The error and the recommended fix was tested with:
- logitechmediaserver-8.3.0-1661147463-arm-linux.tar.gz
- Amazon Linux 2022 running on the AWS Graviton (ARM) processor

**Comment by michaelherger** [comment]
Can you tell me what Perl version (`perl -v`) that system is running? And would adding that dependency actually solve your issues? I'm not familiar with, tbh. And I don't fully understand what your problem is... Can you also get me full output of starting LMS with `-d_startup` added to the command line options?

BTW: I hope you know what you're doing by putting LMS in the cloud. Make sure it's not publicly accessible.

**Comment by saieva** [comment]
Thanks for the Cloud security reminder. I've been running LMS on AWS for over 10 years. I run scans after making changes to confirm it's not publicly accessible.

I'm setting-up LMS on Amazon's new Linux OS (running their ARM-based Graviton processor). The `perl-deprecate` package is not installed by default. When `perl-deprecate` is missing `slimserver.pl` prints a message that `CGI::Cookie` is incompatible, however, the real issue is the version of `CGI::Cookie` can't be determined because `perl-deprecate` is missing. I'm reporting the issue and recommending a potential solution so the error message is accurate in this situation.

Regards,

Sal.

```
ip-192-168-100-246> uname -a
Linux ip-192-168-100-246.ec2.internal 5.15.57-29.131.amzn2022.aarch64 #1 SMP Thu Aug 11 23:18:29 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
ip-192-168-100-246> perl -v

This is perl 5, version 32, subversion 1 (v5.32.1) built for aarch64-linux-thread-multi
(with 51 registered patches, see perl -V for more detail)

Copyright 1987-2021, Larry Wall

Perl may be copied only under the terms of either the Artistic License or the
GNU General Public License, which may be found in the Perl 5 source kit.

Complete documentation for Perl, including FAQ lists, should be found on
this system using "man perl" or "perldoc perl".  If you have access to the
Internet, point your browser at http://www.perl.org/, the Perl Home Page.

ip-192-168-100-246> ./slimserver.pl -d_startup
Found SELinux - setting security context to: texrel_shlib_t for *.so files.
Got @INC containing:
/usr/local/logitechmediaserver-8.3.0
/usr/local/lib64/perl5/5.32
/usr/local/share/perl5/5.32
/usr/lib64/perl5/vendor_perl
/usr/share/perl5/vendor_perl
/usr/lib64/perl5
/usr/share/perl5

Extended @INC to contain:
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/aarch64-linux-thread-multi
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/aarch64-linux-thread-multi/auto
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32.1/aarch64-linux-thread-multi
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32.1/aarch64-linux-thread-multi/auto
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/aarch64-linux-thread-multi
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/aarch64-linux-thread-multi/auto
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/aarch64-linux-thread-multi
/usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32
/usr/local/logitechmediaserver-8.3.0/lib
/usr/local/logitechmediaserver-8.3.0/CPAN
/usr/local/logitechmediaserver-8.3.0
/usr/local/logitechmediaserver-8.3.0
/usr/local/lib64/perl5/5.32
/usr/local/share/perl5/5.32
/usr/lib64/perl5/vendor_perl
/usr/share/perl5/vendor_perl
/usr/lib64/perl5
/usr/share/perl5

Loaded module: [version] ok!
Loaded module: [Time::HiRes] ok!
Loaded module: [DBI] ok!
Loaded module: [EV] ok!
Loaded module: [XML::Parser::Expat] ok!
Loaded module: [HTML::Parser] ok!
Loaded module: [JSON::XS] ok!
Loaded module: [Digest::SHA1] ok!
Loaded module: [YAML::XS] ok!
Loaded module: [Sub::Name] ok!
The following modules are loaded after the first attempt:
	constant.pm => /usr/share/perl5/vendor_perl/constant.pm
	vars.pm => /usr/share/perl5/vars.pm
	B/Op_private.pm => /usr/share/perl5/B/Op_private.pm
	XSLoader.pm => /usr/share/perl5/XSLoader.pm
	Slim/Utils/OS/Unix.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS/Unix.pm
	Digest/SHA1.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Digest/SHA1.pm
	warnings/register.pm => /usr/share/perl5/warnings/register.pm
	lib.pm => /usr/lib64/perl5/lib.pm
	Cwd.pm => /usr/lib64/perl5/vendor_perl/Cwd.pm
	common/sense.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/common/sense.pm
	overloading.pm => /usr/share/perl5/overloading.pm
	Carp.pm => /usr/share/perl5/vendor_perl/Carp.pm
	version/regex.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/version/regex.pm
	Slim/Utils/OS.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS.pm
	version.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/version.pm
	XML/Parser/Expat.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/XML/Parser/Expat.pm
	re.pm => /usr/lib64/perl5/re.pm
	Digest/base.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Digest/base.pm
	Config_heavy.pl => /usr/lib64/perl5/Config_heavy.pl
	strict.pm => /usr/share/perl5/strict.pm
	YAML/XS/LibYAML.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/YAML/XS/LibYAML.pm
	POSIX.pm => /usr/lib64/perl5/POSIX.pm
	Time/HiRes.pm => /usr/lib64/perl5/vendor_perl/Time/HiRes.pm
	YAML/XS.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/YAML/XS.pm
	JSON/XS.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/JSON/XS.pm
	B/Deparse.pm => /usr/share/perl5/B/Deparse.pm
	overload.pm => /usr/share/perl5/overload.pm
	base.pm => /usr/share/perl5/base.pm
	File/Spec.pm => /usr/lib64/perl5/vendor_perl/File/Spec.pm
	Getopt/Long.pm => /usr/share/perl5/vendor_perl/Getopt/Long.pm
	Scalar/Util.pm => /usr/lib64/perl5/vendor_perl/Scalar/Util.pm
	HTML/Entities.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/HTML/Entities.pm
	DynaLoader.pm => /usr/lib64/perl5/DynaLoader.pm
	FindBin.pm => /usr/share/perl5/FindBin.pm
	feature.pm => /usr/share/perl5/feature.pm
	Tie/Hash.pm => /usr/share/perl5/Tie/Hash.pm
	version/vxs.pm => /usr/lib64/perl5/vendor_perl/version/vxs.pm
	File/Spec/Unix.pm => /usr/lib64/perl5/vendor_perl/File/Spec/Unix.pm
	List/Util.pm => /usr/lib64/perl5/vendor_perl/List/Util.pm
	HTML/Parser.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/HTML/Parser.pm
	Exporter/Heavy.pm => /usr/share/perl5/vendor_perl/Exporter/Heavy.pm
	Fcntl.pm => /usr/lib64/perl5/Fcntl.pm
	Config_git.pl => /usr/lib64/perl5/Config_git.pl
	File/Spec/Functions.pm => /usr/lib64/perl5/vendor_perl/File/Spec/Functions.pm
	File/Basename.pm => /usr/share/perl5/File/Basename.pm
	B.pm => /usr/lib64/perl5/B.pm
	Config.pm => /usr/lib64/perl5/Config.pm
	Slim/bootstrap.pm => /usr/local/logitechmediaserver-8.3.0/Slim/bootstrap.pm
	File/Path.pm => /usr/share/perl5/vendor_perl/File/Path.pm
	Sub/Name.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Sub/Name.pm
	EV.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/EV.pm
	DBI.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/DBI.pm
	attributes.pm => /usr/lib64/perl5/attributes.pm
	warnings.pm => /usr/share/perl5/warnings.pm
	Slim/Utils/OS/Linux.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS/Linux.pm
	Slim/Utils/OSDetect.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OSDetect.pm
	Exporter.pm => /usr/share/perl5/vendor_perl/Exporter.pm
	Symbol.pm => /usr/share/perl5/Symbol.pm

The following modules are loaded after the second attempt:
	constant.pm => /usr/share/perl5/vendor_perl/constant.pm
	vars.pm => /usr/share/perl5/vars.pm
	B/Op_private.pm => /usr/share/perl5/B/Op_private.pm
	XSLoader.pm => /usr/share/perl5/XSLoader.pm
	Slim/Utils/OS/Unix.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS/Unix.pm
	Digest/SHA1.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Digest/SHA1.pm
	warnings/register.pm => /usr/share/perl5/warnings/register.pm
	lib.pm => /usr/lib64/perl5/lib.pm
	Cwd.pm => /usr/lib64/perl5/vendor_perl/Cwd.pm
	common/sense.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/common/sense.pm
	overloading.pm => /usr/share/perl5/overloading.pm
	Carp.pm => /usr/share/perl5/vendor_perl/Carp.pm
	version/regex.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/version/regex.pm
	Slim/Utils/OS.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS.pm
	version.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/version.pm
	XML/Parser/Expat.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/XML/Parser/Expat.pm
	re.pm => /usr/lib64/perl5/re.pm
	Digest/base.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Digest/base.pm
	Config_heavy.pl => /usr/lib64/perl5/Config_heavy.pl
	strict.pm => /usr/share/perl5/strict.pm
	YAML/XS/LibYAML.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/YAML/XS/LibYAML.pm
	POSIX.pm => /usr/lib64/perl5/POSIX.pm
	Time/HiRes.pm => /usr/lib64/perl5/vendor_perl/Time/HiRes.pm
	YAML/XS.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/YAML/XS.pm
	JSON/XS.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/JSON/XS.pm
	B/Deparse.pm => /usr/share/perl5/B/Deparse.pm
	overload.pm => /usr/share/perl5/overload.pm
	base.pm => /usr/share/perl5/base.pm
	File/Spec.pm => /usr/lib64/perl5/vendor_perl/File/Spec.pm
	Getopt/Long.pm => /usr/share/perl5/vendor_perl/Getopt/Long.pm
	Scalar/Util.pm => /usr/lib64/perl5/vendor_perl/Scalar/Util.pm
	HTML/Entities.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/HTML/Entities.pm
	DynaLoader.pm => /usr/lib64/perl5/DynaLoader.pm
	FindBin.pm => /usr/share/perl5/FindBin.pm
	feature.pm => /usr/share/perl5/feature.pm
	Tie/Hash.pm => /usr/share/perl5/Tie/Hash.pm
	version/vxs.pm => /usr/lib64/perl5/vendor_perl/version/vxs.pm
	File/Spec/Unix.pm => /usr/lib64/perl5/vendor_perl/File/Spec/Unix.pm
	List/Util.pm => /usr/lib64/perl5/vendor_perl/List/Util.pm
	HTML/Parser.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/HTML/Parser.pm
	Exporter/Heavy.pm => /usr/share/perl5/vendor_perl/Exporter/Heavy.pm
	Fcntl.pm => /usr/lib64/perl5/Fcntl.pm
	Config_git.pl => /usr/lib64/perl5/Config_git.pl
	File/Spec/Functions.pm => /usr/lib64/perl5/vendor_perl/File/Spec/Functions.pm
	File/Basename.pm => /usr/share/perl5/File/Basename.pm
	B.pm => /usr/lib64/perl5/B.pm
	Config.pm => /usr/lib64/perl5/Config.pm
	Slim/bootstrap.pm => /usr/local/logitechmediaserver-8.3.0/Slim/bootstrap.pm
	File/Path.pm => /usr/share/perl5/vendor_perl/File/Path.pm
	Sub/Name.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/Sub/Name.pm
	EV.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/EV.pm
	DBI.pm => /usr/local/logitechmediaserver-8.3.0/CPAN/arch/5.32/DBI.pm
	attributes.pm => /usr/lib64/perl5/attributes.pm
	warnings.pm => /usr/share/perl5/warnings.pm
	Slim/Utils/OS/Linux.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OS/Linux.pm
	Slim/Utils/OSDetect.pm => /usr/local/logitechmediaserver-8.3.0/Slim/Utils/OSDetect.pm
	Exporter.pm => /usr/share/perl5/vendor_perl/Exporter.pm
	Symbol.pm => /usr/share/perl5/Symbol.pm

The following CPAN modules were found but cannot work with Logitech Media Server:
  CGI::Cookie (loaded <not found>, need 1.27)

To fix this problem you have several options:
1. Install the latest version of the module(s) using CPAN: sudo cpan Some::Module
2. Update the module's package using apt-get, yum, etc.
3. Run the .tar.gz version of Logitech Media Server which includes all required CPAN modules.
```

**Comment by kimheino** [comment]
This same bug affects RHEL9.

**Comment by michaelherger** [comment]
Hmm... I thought I had responded here. I'm sorry for the delay.

I see you install from the tarball. Wouldn't the .rpm file work? 

I'm still not sure why that change would be required in your case, as it seems to be working for most users. I _think_ the issue sometimes is when LMS is loading a system library before LMS' own, or when LMS only finds the Perl part of a module, but not its binary. But in your case the lib folder seems to be there towards the top of the list. Wouldn't there be a `/usr/local/logitechmediaserver-8.3.0/lib/CGI/Cookie.pm` file?

**Comment by saieva** [comment]
The issue is a missing `perl-deprecate` module. If it's missing the LMS startup reports an incorrect error loading `CGI::Cookie` (note, the error message says `loaded <not found>`). I believe `CGI::Cookie` is loaded, it just can't determine the version because `perl-deprecate` is missing.

It took me a few hours to troubleshoot so I wanted to share the knowledge and suggest a potential solution.

**Comment by mw9** [comment]
@saieva 

As far as I can tell, the `deprecate` pragma is part of perl core, so I am surprised that there exists a distribution that does not include it. Certainly Debian does. But I have no firm knowledge of this.

- EDIT: Debian includes it in its `perl-modules `package, which forms part of a standard perl installation on that platform.

Here's how it is implemented: https://github.com/Perl/perl5/blob/v5.36.0/lib/deprecate.pm

My gut reaction is that this is a deficiency in your perl installation, not LMS. I may be wrong. Do you know why the Amazon Linux 2022 distribution excludes it ?

As for why you are getting a message:

`CGI::Cookie` had been part of perl core for some time, but was dropped in perl 5.22. In anticipation of that, `CGI::Cookie` made use of the `deprecate` pragma.

I conclude that you must have a slightly "out of date" version of `CGI::Cookie`, somewhere in your perl installation, which still uses this pragma. And, if your perl installation omits the pragma, expect trouble. The current version of `CGI::Cookie` no longer does. Refer commit: https://github.com/leejo/CGI.pm/commit/021f9fdee8e922096dac68f155f574aebb76be1c.

But, LMS does package its own (ancient) version of `CGI::Cookie`, which should be found before any other installed version, and doesn't use the `deprecate` pragma, so you really shouldn't be seeing a message at all. I guess the answer might lie in the LMS bootstrap module. I'll post a follow up if I figure it out.

**Comment by mw9** [comment]
> But, LMS does package its own (ancient) version of `CGI::Cookie`, which should be found before any other installed version, and doesn't use the `deprecate` pragma, so you really shouldn't be seeing a message at all. I guess the answer might lie in the LMS bootstrap module. I'll post a follow up if I figure it out.

Well, no problem with the perl bootstrap module, it prefers the LMS packaged `CGI::Cookie` over a system installed one.

So I think @michaelherger has already posed the key question:
> Wouldn't there be a `/usr/local/logitechmediaserver-8.3.0/lib/CGI/Cookie.pm` file?

**Comment by kimheino** [comment]
My RHEL9 install was clean new install, with system CGI::Cookie (4.51) and bundled lib/CGI/Cookie.pm (1.27). I'm using git snapshot.

**Comment by mw9** [comment]
@kimheino 
Well, that’s quite surprising. Neither `CGI:Cookie` v4.51 nor v1.27 uses the `deprecate` pragma. The OP suggests that, by ensuring the Perl installation implemented the `deprecate` pragma, the error message disappeared. So it would be good to know what is really going on.

The source of the error seems to stem from here:
https://github.com/Logitech/slimserver/blob/public/8.3/Slim/bootstrap.pm#L379

Basically a failure in the preceding `eval "use CGI:Cookie ()";`.

I don’t have access to an RHEL9 system, so not something that I can trace/debug. I think RHEL9 includes Perl v5.32, as does my Debian system, which works as expected.

**Comment by kimheino** [comment]
Yes, installing `perl-deprecate-0.04-479.el9.noarch` fixed this problem for me. RHEL9's perl is 5.32.1.

I'm not a perl programmer so I can't really debug this more. I guess you can replicate this problem with AlmaLinux 9 and other RHEL9 rebuilds.

**Comment by michaelherger** [comment]
I think both of you who encounter this issue aren't running LMS from a package, but from the tarball or a git clone. Please try with the package instead.

**Comment by mw9** [comment]
OK, I have replicated the problem by removing the `deprecate` pragma from an otherwise clean Debian perl.

Turn out that the LMS packaged `CGI::Cookie` depends on the LMS packaged `CGI::Util`. And it is that `CGI::Util` which uses the `deprecate` pragma.

I see that, in first instance, LMS took up `CGI::Util` in its perl 5.22 hierarchy, before moving it to the 'all' hierarchy. It was perl 5.22 that dropped it from core.

Some possible solutions:

1. Do nothing
I don't like this much, because there are clearly distributions that, rightly or wrongly, don't include the `deprecate` pragma by default.

2. Modify the LMS `CGI::Util` to remove the use of the `deprecate` pragma.
This is probably the easiest and least invasive approach. The pragma does nothing whatsoever for us, as it is a no-op unless the module is loaded from the perl core directory. As the module is loaded from the LMS hierarchy, it is a no-op.
3. Upgrade the LMS packaged `CGI::Cookie` and `CGI::Util` to current versions
Not much more work, but there have been other changes in these modules which may or may not introduce changes in behaviour.


I have tested option 2 - it seems to working fine.

**Comment by michaelherger** [comment]
Thank you _very_ much, @mw9! I'll give that a try.

**Comment by mw9** [comment]
I'll keep my fingers crossed. Hopefully @saieva and @kimheino will be able to confirm that the fix applied is, indeed, effective on their systems.

**Comment by kimheino** [comment]
Yes, works for me on RHEL9.

**Comment by michaelherger** [comment]
Thank you for the additional information. Are you running from the tar ball rather than the .rpm file? Have you tried the latter too?

Your log shows that LMS doesn’t load anything from ˋ /usr/local/logitechmediaserver-8.3.0/libˋ - which causes your problem. ˋCGI::Cookiesˋ would be in there. Can you confirm that module is there?

It would be great if you could double check the regular installer. I could imagine some necessary initialisation is missing when running from the source directly.
